dll/win32/iphlpapi/iphlpapi_main.c

```diff
@@ -2620,12 +2622,93 @@ SetIpForwardEntryToStack(PMIB_IPFORWARDROW pRoute)
 }
 
 DWORD WINAPI
-NhGetInterfaceNameFromDeviceGuid(DWORD dwUnknown1,
-                                 DWORD dwUnknown2,
-                                 DWORD dwUnknown3,
+NhGetInterfaceNameFromDeviceGuid(GUID * pInterfaceGUID,
+                                 PWCHAR pInterfaceName,
+                                 PULONG pOutBufLen,
                                  DWORD dwUnknown4,
                                  DWORD dwUnknown5)
 {
-    FIXME("NhGetInterfaceNameFromDeviceGuid() stub\n");
-    return 0L;
+    LPOLESTR pStr;
+    WCHAR szKeyName[2*MAX_PATH];
+    HKEY hKey;
+    DWORD cbData = 0, type = REG_SZ;
+
+    if (StringFromCLSID(pInterfaceGUID, &pStr) != ERROR_SUCCESS)
+    {
+        // failed to convert guid to string
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    StringCbPrintfW(szKeyName, sizeof(szKeyName), L"SYSTEM\\CurrentControlSet\\Control\\Network\\{4D36E972-E325-11CE-BFC1-08002BE10318}\\%s\\Connection", pStr);
+    CoTaskMemFree(pStr);
+
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, szKeyName, 0, KEY_READ, &hKey) != ERROR_SUCCESS) 
+    {
+        // failed to find adapter entry
+        return ERROR_FILE_NOT_FOUND;
+    }
+
+    if (RegQueryValueExW(hKey, L"Name", NULL, &type, NULL, &cbData) != ERROR_SUCCESS)
+    {
+        // failed to read adapter name
+        RegCloseKey(hKey);
+        return ERROR_NO_DATA;
+    }
+
+    if (cbData > *pOutBufLen)
+    {
+        // buffer size too small
+        RegCloseKey(hKey);
+        return ERROR_BUFFER_OVERFLOW;
+    }
+
+    RegQueryValueExW(hKey, L"Name", NULL, &type, (PVOID)pInterfaceName, pOutBufLen);
+    RegCloseKey(hKey);
+    return ERROR_SUCCESS;
+}
+
+DWORD WINAPI
+NhGetInterfaceNameFromGuid(GUID * pInterfaceGUID,
+                           PWCHAR pInterfaceName,
+                           PULONG pOutBufLen,
+                           DWORD dwUnknown4,
+                           DWORD dwUnknown5)
+{
+    LPOLESTR pStr;
+    WCHAR szKeyName[2*MAX_PATH];
+    HKEY hKey;
+    DWORD cbData = 0, type = REG_SZ;
```

Comment:
Why initialize `type`?
It's an out for `RegQueryValueExW`.

Reply:
Oh, you're right.